src/rdkafka_cgrp.c

```diff
@@ -2535,22 +3013,38 @@ rd_kafka_cgrp_incremental_assign (rd_kafka_cgrp_t *rkcg,
         rkcg->rkcg_c.assignment_size = rkcg->rkcg_assignment->cnt;
         rd_kafka_wrunlock(rkcg->rkcg_rk);
 
-        rd_kafka_cgrp_set_join_state(rkcg,
-                                     RD_KAFKA_CGRP_JOIN_STATE_ASSIGNED);
+        rd_kafka_cgrp_set_join_state(rkcg, RD_KAFKA_CGRP_JOIN_STATE_ASSIGNED);
 
         if (RD_KAFKA_CGRP_CAN_FETCH_START(rkcg))
                 rd_kafka_cgrp_partitions_fetch_start(
                         rkcg, partitions, 0);
 
+check_rejoin:
+
+        if (rkcg->rkcg_rebalance_rejoin) {
+                rd_kafka_dbg(rkcg->rkcg_rk,
+                             CGRP|RD_KAFKA_DBG_CONSUMER, "CGRP",
+                             "Group \"%s\": rejoining group to redistribute "
+                             "previously owned partitions",
+                             rkcg->rkcg_group_id->str);
+                rd_kafka_cgrp_set_join_state(rkcg,
```

Comment:
that's been renamed `rd_kafka_cgrp_revoke_rejoin` and does stuff we don't want to do here.

Reply:
I've added the function, but i don't get the mental framework, and based on what I know i'm not sure i agree it's better. if this were a great abstraction, yes, but these state transitions are very intermingled with understanding the code, so i just see this as a poor attempt at creating an abstraction that doesn't succeed in making thinking about things easier, more obfuscates things because it put something you want to know when reading the code behind a layer of indirection.

on the subject of maintainability, it would be dramatically easier to maintain this type of logic in Task based architecture... communication between threads is not your concern. what threads stuff runs on is not your concern. you don't explicitly write this state transition logic. you have a nice object representing the async workflow etc. dramatically easier.

Reply:
note: you answered my rant here later.